### DISCUSSION TITLE
ctime & mtime date format

### DISCUSSION THREAD

**Discussion by ghost** [sxyazi/yazi#1643] [5. Polls]
### `yazi --debug` output

```Shell
Yazi
    Version: 0.3.3 (Arch Linux 2024-09-05)
    Debug  : false
    OS     : linux-x86_64 (unix)

Ya
    Version: 0.3.3 (Arch Linux 2024-09-05)

Emulator
    Emulator.via_env: ("foot", "")
    Emulator.via_csi: Ok(Foot)
    Emulator.detect : Foot

Adapter
    Adapter.matches: Sixel

Desktop
    XDG_SESSION_TYPE           : Some("wayland")
    WAYLAND_DISPLAY            : Some("wayland-1")
    DISPLAY                    : Some(":0")
    SWAYSOCK                   : None
    HYPRLAND_INSTANCE_SIGNATURE: Some("0f594732b063a90d44df8c5d402d658f27471dfe_1726362800_736104487")
    WAYFIRE_SOCKET             : None

SSH
    shared.in_ssh_connection: false

WSL
    WSL: false

Variables
    SHELL              : Some("/usr/bin/zsh")
    EDITOR             : Some("helix")
    VISUAL             : Some("helix")
    YAZI_FILE_ONE      : None
    YAZI_CONFIG_HOME   : None

Text Opener
    default: Some(Opener { run: "helix \"$@\"", block: true, orphan: false, desc: "$EDITOR", for_: None, spread: true })
    block  : Some(Opener { run: "helix \"$@\"", block: true, orphan: false, desc: "$EDITOR", for_: None, spread: true })

Multiplexers
    TMUX               : false
    tmux version       : No such file or directory (os error 2)
    ZELLIJ_SESSION_NAME: None
    Zellij version     : No such file or directory (os error 2)

Dependencies
    file             : 5.45
    ueberzugpp       : No such file or directory (os error 2)
    ffmpegthumbnailer: 2.2.2
    magick           : 7.1.1-38
    fzf              : 0.54.3
    fd               : No such file or directory (os error 2)
    rg               : 14.1.1
    chafa            : No such file or directory (os error 2)
    zoxide           : 0.9.4
    7z               : 17.05
    7zz              : No such file or directory (os error 2)
    jq               : 1.7.1


--------------------------------------------------
When reporting a bug, please also upload the `yazi.log` log file - only upload the most recent content by time.
You can find it in the "/home/alec/.local/state/yazi" directory.
```

### Please describe the problem you're trying to solve

The current date format for ctime and mtime linemodes are mm/dd and I wish to be able to change it to the more conventional dd/mm without writing a custom linemode.

Anyway, thank you for making this super fast ducky file manager. Cheers!

### Would you be willing to contribute this feature?

- [ ] Yes, I'll give it a shot

### Describe the solution you'd like

Maybe it would be good to have an option to format linemodes in the configuration file?

I would contribute but all I know is some terrible python, sorry.

### Validations

- [X] I have searched the existing issues
- [X] The latest nightly build of Yazi doesn't already have this feature

**Comment by sxyazi** [comment]
> Maybe it would be good to have an option to format linemodes in the configuration file?

This kind of ruins the point of allowing users to override the linemode, because the whole purpose of supporting it was to avoid adding extra options. I'm certain that it's impossible to cover all users' needs with options, and doing so would bloat Yazi. Plus, overriding it is really simple, requiring just a few lines of Lua code, and it has no impact on performance or the need to recompile the source code.

However, I can consider changing the default date format from `%m/%d` to `%d/%m` if it makes things more convenient for most of users - I'll turn this issue into a discussion to see what other users think.

**Reply by xfzv** [reply]
> However, I can consider changing the default date format from `%m/%d` to `%d/%m` if it makes things more convenient for most of users - I'll turn this issue into a discussion to see what other users think.

My two cents: I'm using a custom linemode so I don't really care about the default. `%m/%d` makes much more sense to me than `%d/%m` though.

Assuming chronological listing is wanted, let's take March 8th and August 3rd (same year) for example:

- With `%m/%d`:
    - 03/08 (March 8th)
    - 08/03 (August 3rd)

- With `%d/%m`:
    - 03/08 (August 3rd)
    - 08/03 (March 8th)

August 3rd being listed before March 8th just feels wrong to me.

**Comment by verga** [comment]
I advice using the international standard [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html)
```
YYYY-MM-DD hh:mm
```
or other combinations [https://en.wikipedia.org/wiki/ISO_8601](https://en.wikipedia.org/wiki/ISO_8601)

**Reply by prosoitos** [reply]
100% agree. That's the default in programming languages and most tools out there

**Reply by apprenti-sorcier** [reply]
Agreed. That option is the middle ground that would satisfy most users and it should be the default.

**Comment by njhallett** [comment]
or maybe use the locale of the user's system

```sh
locale -k LC_TIME
```

```ps1
([cultureinfo]::CurrentCulture.DateTimeFormat).ShortDatePattern
```

**Comment by jozsefk9** [comment]
`dd/mm` would work better for me too. Didn't get used to `mm/dd` at all. But if I could customize it by myself, or if I would know how to do it, this is how I would make it to look like: `03Nov2024 @ 22^59^47`

**Reply by ghost** [reply]
I solved this long ago by making custom linemodes in the `init.lua` file.

```lua
function Linemode:mtimev2()
	local time = math.floor(self._file.cha.modified or 0)
		time = time and os.date("%g-%m-%d %H:%M", time) or ""
	return ui.Line(string.format(time))
end

function Linemode:ctimev2()
	local time = math.floor(self._file.cha.created or 0)
		time = time and os.date("%g-%m-%d %H:%M", time) or ""
	return ui.Line(string.format(time))
end

``` 

and then if you like set one of those to the default linemode in `~/.config/yazi/yazi.toml`(and add those to the keybinds too).

```toml
[manager]
linemode = "whatever-mode-you-like"
```

Note: I took this and edited from the example for custom linemodes [here](https://yazi-rs.github.io/docs/configuration/yazi/#manager.linemode).

**Reply by jozsefk9** [reply]
Great! Thanks so much! It's working.

**Reply by moshpirit** [reply]
> I solved this long ago by making custom linemodes in the `init.lua` file.

How exactly do you tell Yazi to read the package? I tried with this, but it doesn't seem to work:

```lua
[plugins]
preload = [
    "~/.config/yazi/plugins/timemodes.yazi/init.lua",
    "~/.config/yazi/init.lua",
]
```

I created it both as a local init.lua and as a plugin, but it doesn't seem to be loaded

**Reply by ghost** [reply]
It is not necesary to do that. Did you set the corresponding keybinds and/or default linemode?

I have my custom linemodes keybinds set [here](https://codeberg.org/299792458ms/dotfiles/src/branch/main/dot_config/yazi/keymap.toml#L77-L78). If those are not explicitly set in `keymap.toml` (overwrite/append) yazi will fall back to the defaults.

**Reply by moshpirit** [reply]
Thanks! What a wonderful tool! I turned it into this: if the file/directory is from previous years, it will be shown, if not, will stay hidden to keep it simple:

```lua
function Linemode:mtimev2()
    local current_year = os.date("%Y") -- Get the current year as a string
    local time = math.floor(self._file.cha.modified or 0)
    if time > 0 then
        local file_year = os.date("%Y", time) -- Get the file's year
        local time_format
        if file_year == current_year then
            time_format = os.date("%d/%b %H:%M", time) -- Without year
        else
            time_format = os.date("%d/%b/%Y %H:%M", time) -- With year
        end
        return ui.Line(string.format(time_format))
    end
    return ui.Line("") -- Return empty if no time exists
end

function Linemode:ctimev2()
    local current_year = os.date("%Y") -- Get the current year as a string
    local time = math.floor(self._file.cha.created or 0)
    if time > 0 then
        local file_year = os.date("%Y", time) -- Get the file's year
        local time_format
        if file_year == current_year then
            time_format = os.date("%d/%b %H:%M", time) -- Without year
        else
            time_format = os.date("%d/%b/%Y %H:%M", time) -- With year
        end
        return ui.Line(string.format(time_format))
    end
    return ui.Line("") -- Return empty if no time exists
end
```

**Reply by matejdro** [reply]
Is this compatible with the latest Yazi? For me `time` is always `0`

**Reply by rod-stuchi** [reply]
It's very simple to override the methods, check out how it's done:
https://github.com/sxyazi/yazi/blob/main/yazi-plugin/preset/components/linemode.lua

init.lua
```lua
function Linemode:mtime()
  -- your override logic
end

**Reply by dmolin** [reply]
It would be much simpler to use the current system locale instead of having to write Lua code to tweak stuff. 
Writing a Lua function for the linemode doesn't fix things; Yazi will still use the YY/MM/DD everywhere else. 

Picking up the system locale (LC_TIME) would be the most efficient thing to do, so each user can set the date format he wants to and this will apply to Yazi everywhere dates are involved.

**Reply by barsikus007** [reply]
I made my own init.lua based on original, but with 2077-12-31 format
https://github.com/sxyazi/yazi/blob/157156b5b8f36db15b2ba425c7d15589039a9e1e/yazi-plugin/preset/components/linemode.lua#L25

```lua
function strip_date_year(time_to_format)
  local time = math.floor(time_to_format or 0)
  if time == 0 then
    return ""
  elseif os.date("%Y", time) == os.date("%Y") then
    return os.date("%m-%d %H:%M", time)
  else
    return os.date("%Y-%m-%d", time)
  end
end
function Linemode:btime()
  return strip_date_year(self._file.cha.btime)
end
function Linemode:mtime()
  return strip_date_year(self._file.cha.mtime)
end
```

**Comment by Kapparina** [comment]
I agree that either utilising ISO 8601 yyyy-MM-dd (maybe also HH:mm:ss) or the system locale date format is better than what's currently presented to the user.

**Comment by Tekks97** [comment]
Would love to see that option implemented as well.

**Comment by alberti42** [comment]
I was wondering whether someone worked out a PR for this missing feature. Probably, honoring
```
locale -k LC_TIME
```
is the best way to go. Otherwise, Yazi could offer a format for localizing date and time strings. Before coding a PR, it would be relevant to hear the thoughts of @sxyazi 

> This kind of ruins the point of allowing users to override the linemode, because the whole purpose of supporting it was to avoid adding extra options. I'm certain that it's impossible to cover all users' needs with options, and doing so would bloat Yazi. 

I understand this argument. However, the `linemode` is not the same as when sorting files per modification / birth time. I use it a lot in large directories to find out the last files I was working with. Why simply not honor the locale of the terminal? This would not even add an extra configuration.

Alternatively, why not offer a similar customization like `linemode` for `timemode` and `sizemode`? In this way, the user can choose how to display time and size the way they want. This probably would be the most yaziest way to go and mirror what already exists with `linemode`.

**Reply by sxyazi** [reply]
It's a good idea to respect the user's system default date/time format, but Lua's `os.date()` doesn't seem to have the flexibility to create a datetime string that matches the current linemode design: 

- if a file is from this year, shows month, day, hour, and minute
- if it's not from this year, shows year, month, and day

In JavaScript I can do it like this, respectively:

```javascript
new Date().toLocaleTimeString([], { month: 'numeric', day: 'numeric', hour: '2-digit', minute: '2-digit' })
'2/11, 05:47 PM'
```

```javascript
new Date().toLocaleDateString([], { year: 'numeric', month: 'numeric', day: 'numeric' })
'2/11/2026'
```

Do you know any way to do the same in Lua?

**Reply by alberti42** [reply]
I did some research on this — asked Claude and then dug into the Lua docs to understand it better. I found this in the https://www.lua.org/manual/5.3/manual.html#pdf-os.date:

>  If format is not "*t", then date returns the date as a string, formatted according to the same rules as the ISO C function strftime.

And also:

>  When called without arguments, date returns a reasonable date and time representation that depends on the host system and on the current locale.

  So it turns out os.date does support locale through C's strftime specifiers:

  - %c — full locale date and time (what os.date() with no args uses)
  - %x — locale date
  - %X — locale time

Meaning os.date("%x", time) would already give you a locale-respecting date string, no Rust changes needed. The catch is that these are preset bundles: %x gives you whatever the locale defines as "a date," typically including the year. You can't say "give me month + day + hour + minute in locale order but without the year."

So if we're okay with showing the full locale date for all files, os.date("%x") already works today in Lua. Something like:

```lua
function Linemode:size_and_mtime()
  local time = math.floor(self._file.cha.mtime or 0)
  if time == 0 then
    time = ""
  else
    time = os.date("%x %X", time)
  end

  local size = self._file:size()
  return string.format("%s %s", size and ya.readable_size(size) or "", time)
end
```

If you want finer control though, like in your JavaScript snippet, where you pick which components to show and the locale decides the ordering, then probably the best way would be to expose a small Rust function to Lua for that.

Rust has the https://docs.rs/chrono crate (already used by yazi, I believe) and could use https://docs.rs/icu or https://docs.rs/pure-rust-locales for locale-aware formatting. The approach would be:

1. Rust side: Expose a function to Lua (via mlua) like ya.format_datetime(timestamp, components) where components specifies which fields to include (year, month, day, hour, minute), and the function returns a string formatted according to LC_TIME.
2. Lua side: Users keep full control. The default linemode would call something like:

```lua
function Linemode:size_and_mtime()
  local time = math.floor(self._file.cha.mtime or 0)
  if time == 0 then
    time = ""
  else
    time = ya.format_datetime(time, { "month", "day", "hour", "minute" })
  end

  local size = self._file:size()
  return string.format("%s %s", size and ya.readable_size(size) or "", time)
end
```

This would fit sxyazi's philosophy:
- No new config options → no bloat
- Lua stays in control → users can still override everything
- Locale is respected → the Rust function handles LC_TIME ordering
- Solves the JS gap → it's the Lua equivalent of toLocaleTimeString([], { month: 'numeric', ... })

**Reply by alberti42** [reply]
I looked more into it and found the perfect solution for my needs. I believe, with suitable customization, this solution will work for anybody else. I don't think there is any need for change in Yazi code.

Yazi is using by default these keybindings:

```toml
	# Sorting
	{ on = [ ",", "m" ], run = [ "sort mtime --reverse=no", "linemode mtime" ],  desc = "Sort by modified time" },
	{ on = [ ",", "M" ], run = [ "sort mtime --reverse=yes", "linemode mtime" ], desc = "Sort by modified time (reverse)" },
	{ on = [ ",", "b" ], run = [ "sort btime --reverse=no", "linemode btime" ],  desc = "Sort by birth time" },
	{ on = [ ",", "B" ], run = [ "sort btime --reverse=yes", "linemode btime" ], desc = "Sort by birth time (reverse)" },
	{ on = [ ",", "e" ], run = "sort extension --reverse=no",                    desc = "Sort by extension" },
	{ on = [ ",", "E" ], run = "sort extension --reverse=yes",                   desc = "Sort by extension (reverse)" },
	{ on = [ ",", "a" ], run = "sort alphabetical --reverse=no",                 desc = "Sort alphabetically" },
	{ on = [ ",", "A" ], run = "sort alphabetical --reverse=yes",                desc = "Sort alphabetically (reverse)" },
	{ on = [ ",", "n" ], run = "sort natural --reverse=no",                      desc = "Sort naturally" },
	{ on = [ ",", "N" ], run = "sort natural --reverse=yes",                     desc = "Sort naturally (reverse)" },
	{ on = [ ",", "s" ], run = [ "sort size --reverse=no", "linemode size" ],    desc = "Sort by size" },
	{ on = [ ",", "S" ], run = [ "sort size --reverse=yes", "linemode size" ],   desc = "Sort by size (reverse)" },
	{ on = [ ",", "r" ], run = "sort random --reverse=no",                       desc = "Sort randomly" },
```

I did not need to change them, but this was for me the key to understand what was happening. When you go in sort mode, say sorting by size, it does two things: it switches to sorting by size and it changes the linemode to `size`.

To customize the date the way I wanted it for me, all what I had to do was to add to my `init.lua`:

```lua
function Linemode:size_and_mtime()
  local time = math.floor(self._file.cha.mtime or 0)
  local time_str
  if time == 0 then
    time_str = ""
  else
    time_str = os.date("%x %H:%M", time)
  end

  local size = self._file:size()
  local size_str = size and ya.readable_size(size) or ""
  if size then
    return string.format("%s %s", size_str, time_str)
  else
    return string.format("%s", time_str)
  end
end

function Linemode:mtime()
  local time = math.floor(self._file.cha.mtime or 0)
  local time_str
  if time == 0 then
    time_str = ""
  else
    time_str = os.date("%x %H:%M", time)
  end
  return time_str
end

function Linemode:btime()
  local time = math.floor(self._file.cha.btime or 0)
  local time_str
  if time == 0 then
    time_str = ""
  else
    time_str = os.date("%x %H:%M", time)
  end
  return time_str
end

function Linemode:my_default()
  return self:size_and_mtime()
end
```

Please check the description in https://www.lua.org/manual/5.3/manual.html#pdf-os.date if you want to customize the date to a different format that works best for you.

I also added to my keybindgs:

```toml
[[mgr.prepend_keymap]]
on   = [ "m", "d" ]
run  = [
  "linemode my_default",
  ]
desc = "Linemode: my own default"
```

and use linemode `my_default` as a wrapper for `size_and_mtime`. This makes it easier in the future to switch to a different default `linemode`, if `I decide I want to change. In the `yazi.toml`, you may want to add:
```toml
[mgr]
linemode       = "my_default"
```
so that `yazi` starts directly with your default linemode.

I hope this can help someone else in the future.

@sxyazi: thanks a lot for the prompt reply and developing this wonderful tool.

**Comment by apprenti-sorcier** [comment]
In case anyone is interested, I added this in yazi’s init.lua (thanks to ChatGPT):

```
-- file size with thousand separators and precise bytes (to distinguish files with similar sizes), no ambiguous units (like capital K): see https://en.wikipedia.org/wiki/Byte#Multiple-byte_units
function ya.readable_size(size) 
	local s = tostring(math.floor(size))
	return s:reverse():gsub("(%d%d%d)", "%1 "):reverse():gsub("^ ", "")
end

local weekdays = { "Sun.", "Mon.", "Tue.", "Wed.", "Thu.", "Fri.", "Sat." }

-- if file timestamp from today, HH:MM:SS
-- if older than this year, use a ISO-8601 inspired timestamp (T is replaced by a space for easy reading): https://en.wikipedia.org/wiki/ISO_8601
-- to save space, as permitted by the standard, the year can be skipped with a dash
-- also, if within the last week, explicit the days: yesterday or the day of the week

function Linemode:mtime()
	local time = self._file.cha.mtime
	if not time or time == 0 then
		return ""
	end

	time = math.floor(time)

	local t = os.date("*t", time)
	local now = os.date("*t")

	if t.year == now.year then
		local diff = now.yday - t.yday

		if diff == 0 then
			return os.date("%H:%M:%S", time)
		elseif diff == 1 then
			return "yest. " .. os.date("%H:%M:%S", time)
		elseif diff >= 2 and diff <= 7 then
			return weekdays[t.wday] .. " " .. os.date("%H:%M:%S", time)
		else
			return os.date("--%m-%d %H:%M:%S", time)
		end
	else
		return os.date("%Y-%m-%d %H:%M:%S", time)
	end
end
```

Adapt to your preference. Add a similar btime.

P.S. It doesn’t detect the last days if the year changed, but this is a negligible inconvenience, easily solved if absolutely necessary.
